# Accessing and exploration of the AACT database

In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

In [2]:
import pandas as pd
import os
import requests
from sqlalchemy import create_engine

## Accessing AACT database

In [3]:
# Connect to database
aact_host=os.environ["aact_host"]
aact_port=os.environ["aact_port"]
db_name=os.environ["db_name"]
aact_username=os.environ["aact_username"]
aact_password=os.environ["aact_password"]

db_string = "postgres://{}:{}@{}:{}/{}". \
    format(aact_username, aact_password, aact_host, aact_port, db_name)

db = create_engine(db_string)

conn = db.connect()

In [4]:
# Perform count of studies
q = "select count(*) from studies;"

conn.execute(q).fetchall()

[(306660,)]

## Explore data relevant to patients
The AACT database contains ___ tables with ___ data. In order to 
We'll be extracting important information that will be important to patients. I wanted to prioritize which sections will be most important for patients for understanding whether they will be eligible for a trial. We concentrated on information such as the active participation period, the follow-up period, and eligibility criteria.



We designed the categories from our manual analysis of 100 sample trial descriptions. The categories include the information most vital to patients’ understanding of a trial: active participation period, follow up period, and eligibility criteria, to name a few.

Next we labelled each category with its datatype, and tagged each with keywords that appear in or near the information we want.

To assess the validity of the keywords we chose, we conducted frequency evaluations of each indicator. Here’s a sample of what the output looks like:

In [5]:
# Perform selection of brief_summary
condition = "carcinoma, hepatocellular"

q = """SELECT studies.nct_id,
            e.gender,
            e.minimum_age,
            e.maximum_age,
            e.healthy_volunteers,
            e.population,
            e.criteria AS eligibility_criteria,
            bs.description AS brief_summary,
            
            c.downcase_name AS condition
        FROM studies
        FULL OUTER JOIN brief_summaries bs
            ON studies.nct_id=bs.nct_id
        FULL OUTER JOIN eligibilities e
            ON studies.nct_id=e.nct_id
        INNER JOIN conditions c
            ON studies.nct_id=c.nct_id
        INNER JOIN browse_conditions bc
            ON studies.nct_id=bc.nct_id
        WHERE c.downcase_name='{}'
        LIMIT 5;
        """.format(condition)

r = conn.execute(q).fetchall()
col_names = conn.execute(q).keys()

pd.DataFrame(r, columns=col_names)

nct_id gender minimum_age maximum_age healthy_volunteers population  \
0  NCT02847767    All    18 Years         N/A                 No              
1  NCT02847767    All    18 Years         N/A                 No              
2  NCT02794337    All    18 Years         N/A                 No              
3  NCT02794337    All    18 Years         N/A                 No              
4  NCT01605734    All    18 Years    75 Years                 No              

                                eligibility_criteria  \
0  \n        Inclusion Criteria:\n\n        All t...   
1  \n        Inclusion Criteria:\n\n        All t...   
2  \n        Inclusion Criteria:\n\n          -  ...   
3  \n        Inclusion Criteria:\n\n          -  ...   
4  \n        Inclusion Criteria:\n\n          -  ...   

                                       brief_summary  \
0  \n      The high dose per fraction (>10Gy/frac...   
1  \n      The high dose per fraction (>10Gy/frac...   
2  \n      Vast majority of patients with hepatoc...   
3  \n      Vast majority of patients with hepatoc...   
4  \n      TACE is widely used in patients with u...   

                   condition  
0  carcinoma, hepatocellular  
1  carcinoma, hepatocellular  
2  carcinoma, hepatocellular  
3  carcinoma, hepatocellular  
4  carcinoma, hepatocellular

In [6]:
# Identify total number of trials for all with condition
q = """SELECT count(DISTINCT studies.nct_id)
        FROM studies
        FULL OUTER JOIN brief_summaries bs
            ON studies.nct_id=bs.nct_id
        FULL OUTER JOIN eligibilities e
            ON studies.nct_id=e.nct_id
        INNER JOIN conditions c
            ON studies.nct_id=c.nct_id
        INNER JOIN browse_conditions bc
            ON studies.nct_id=bc.nct_id
        WHERE c.downcase_name='{}';
        """.format(condition)

r = conn.execute(q).fetchall()
r[0][0]


222

In [7]:
# Extract information from all HCC trials
# Perform selection of brief_summary
condition = "carcinoma, hepatocellular"

q = """SELECT studies.nct_id,
            e.gender,
            e.minimum_age,
            e.maximum_age,
            e.healthy_volunteers,
            e.population,
            e.criteria AS eligibility_criteria,
            bs.description AS brief_summary,
            c.downcase_name AS condition
        FROM studies
        FULL OUTER JOIN brief_summaries bs
            ON studies.nct_id=bs.nct_id
        FULL OUTER JOIN eligibilities e
            ON studies.nct_id=e.nct_id
        INNER JOIN conditions c
            ON studies.nct_id=c.nct_id
        INNER JOIN browse_conditions bc
            ON studies.nct_id=bc.nct_id
        WHERE c.downcase_name='{}';
        """.format(condition)

r = conn.execute(q).fetchall()
col_names = conn.execute(q).keys()

df = pd.DataFrame(r, columns=col_names)

# Drop duplicate rows
df.drop_duplicates(inplace=True)

df 

nct_id gender minimum_age maximum_age healthy_volunteers  \
0    NCT02847767    All    18 Years         N/A                 No   
2    NCT02794337    All    18 Years         N/A                 No   
4    NCT01605734    All    18 Years    75 Years                 No   
6    NCT01594125    All    20 Years         N/A                 No   
8    NCT01668134    All    18 Years         N/A                 No   
11   NCT01556815    All    18 Years    80 Years                 No   
22   NCT01078415    All    18 Years         N/A                 No   
24   NCT02878109    All    18 Years         N/A                 No   
26   NCT02847468    All    18 Years    99 Years                 No   
28   NCT02115373    All    18 Years         N/A                 No   
30   NCT03941873    All    18 Years         N/A                 No   
31   NCT01040559    All    18 Years         N/A                 No   
33   NCT00987935    All    18 Years         N/A                 No   
35   NCT03766555    All    18 Years         N/A                 No   
37   NCT03418922    All    20 Years         N/A                 No   
39   NCT03191123    All    20 Years    75 Years                 No   
41   NCT03168152    All    18 Years         N/A                 No   
43   NCT02308319    All    18 Years         N/A                 No   
46   NCT01327521    All    18 Years         N/A                 No   
48   NCT02973204    All    18 Years         N/A                 No   
51   NCT01306058    All    18 Years    99 Years                 No   
59   NCT02848638    All    18 Years    85 Years                 No   
61   NCT01411436    All         N/A         N/A                 No   
63   NCT02576964    All    18 Years    75 Years                 No   
65   NCT02403544    All    18 Years    65 Years                 No   
67   NCT01915602    All    18 Years         N/A                 No   
69   NCT01915589    All    18 Years         N/A                 No   
71   NCT03291379    All    18 Years         N/A                 No   
73   NCT02069041    All    20 Years    75 Years                 No   
75   NCT02715362    All    18 Years    69 Years                 No   
..           ...    ...         ...         ...                ...   
510  NCT00116454    All    18 Years    75 Years                 No   
514  NCT00662246    All    18 Years         N/A                 No   
516  NCT00920192    All    18 Years         N/A                 No   
518  NCT00108953    All    18 Years         N/A                 No   
520  NCT00105443    All    18 Years         N/A                 No   
522  NCT00082082    All    18 Years         N/A                 No   
524  NCT00081900    All    18 Years         N/A                 No   
526  NCT00080236    All    18 Years         N/A                 No   
528  NCT00703365    All    18 Years         N/A                 No   
530  NCT00919009    All    20 Years         N/A                 No   
532  NCT00701168    All    19 Years         N/A                      
534  NCT00855218    All    18 Years         N/A                 No   
536  NCT00044512    All    18 Years         N/A                 No   
538  NCT00034333    All    18 Years    80 Years                 No   
540  NCT00936689    All    18 Years         N/A                 No   
542  NCT00687778    All    18 Years    90 Years                 No   
550  NCT00699374    All    18 Years         N/A                 No   
552  NCT00056992    All         N/A         N/A                 No   
554  NCT00763776    All    18 Years         N/A                 No   
558  NCT00247728    All    18 Years    75 Years                 No   
560  NCT00122876    All    18 Years         N/A                 No   
563  NCT00901901    All    18 Years         N/A                 No   
565  NCT00582400    All    18 Years         N/A                 No   
567  NCT00563095    All    18 Years         N/A                      
569  NCT03808766    All    19 Years         N/A                 No   
571  NCT00437424    

## Process text

In [8]:
def clean_text(text):
    
    elements_to_remove = [':',
                           '-',
                           '\n',
                          '\t',
                          ]
    for element in elements_to_remove:
        text = text.replace(element, '')
    
    return text.lower()
    
df.eligibility_criteria.apply(clean_text)[0]

"        inclusion criteria        all the following criteria must be met          1. age > 18 years old          2. multiphase ct scan and/or mri of the liver within 8 weeks of radiation planning             demonstrating                 liver tumours < 5 cm                 no more than 2 discrete liver tumours                 normal liver > 700 cc          3. patients must have hcc diagnosed by either i) pathological confirmation, or ii)             intrahepatic vascular enhancement of the lesion demonstrated by at least two imaging             modalities, or iii) intrahepatic vascular enhancement of the lesion demonstrated by             one imaging modality if afp > 200 in the setting of liver cirrhosis or chronic             hepatitis b without cirrhosis (easl consensus guidelines [2])          4. liver hcc must be deemed unresectable as determined by an experienced hepatobiliary             surgeon, or the patient must be medically inoperable or refuse surgery,          5. patien

## Tokenize words and take frequencies using nltk

In [9]:
from nltk.tokenize import word_tokenize
from nltk import bigrams
from nltk.corpus import stopwords
import matplotlib.pyplot as plt

In [10]:
def remove_stopwords(text, stopwords):
    return [word for word in text if word not in stopwords]

stop_words = set(stopwords.words('english'))

df.eligibility_criteria.apply(clean_text).apply(word_tokenize).apply(remove_stopwords, stopwords=stop_words)

0      [inclusion, criteria, following, criteria, mus...
2      [inclusion, criteria, diagnosis, hcc, ., tissu...
4      [inclusion, criteria, patients, newly, diagnos...
6      [inclusion, criteria, 1., histologically/cytol...
8      [inclusion, criteria, age, ≥, 18, karnofsky, p...
11     [inclusion, criteria, patients, newly, diagnos...
22     [inclusion, criteria, hcc, diagnosed, positive...
24     [inclusion, criteria, patients, least, 18, yea...
26     [inclusion, criteria, patient, older, 18, year...
28     [inclusion, criteria, histologically, confirme...
30     [inclusion, criteriaeach, patient, eligible, p...
31     [inclusion, criteria, hepatocellular, carcinom...
33     [inclusion, criteria, 1., hepatocellular, carc...
35     [inclusion, criteria, age, >, 18, years, hcc, ...
37     [inclusion, criteria, participants, must, conf...
39     [inclusion, criteria, clinical, diagnosis, hep...
41     [inclusion, criteria, patients, hepatocellular...
43     [inclusion, criteria, he

In [11]:
bigrams = df['eligibility_criteria'].apply(clean_text). \
    apply(word_tokenize).  \
    apply(remove_stopwords, stopwords=stop_words). \
    apply(lambda row: list(bigrams(row))). \
    apply(pd.Series).stack().reset_index(drop=True) # Turn series of lists into stacked series

In [ ]:
# Bigram frequencies
val_counts = bigrams.value_counts()[:100]

plt.figure(figsize=(12,6))
plt.bar(val_counts.index.map(str), val_counts.values);

In [15]:
bigrams.value_counts()[6:100]

(., patients)                 121
(4, weeks)                    120
(performance, status)         102
(6, months)                   102
(18, years)                    85
(., exclusion)                 78
(life, expectancy)             77
(criteria, 1.)                 77
((, ))                         75
(patients, must)               73
(3, months)                    69
(dose, study)                  67
(criteria, patients)           66
(first, dose)                  66
(days, prior)                  65
(within, 4)                    64
(least, one)                   62
(upper, limit)                 62
(limit, normal)                59
(written, informed)            59
(oncology, group)              59
(study, drug)                  58
(prior, study)                 58
(childpugh, class)             58
(eastern, cooperative)         57
(cooperative, oncology)        57
(weeks, prior)                 56
(portal, vein)                 54
(therapy, ,)                   53
(study, entry)

Analysis of the top most frequent bigrams from eligibilites of 222 HCC trials showed common features from these trials. These bigrams can now be used to identify general wording structures. For instance, lengths of time (eg, "2 weeks" or "4 weeks") or can be extracted from these

- 4 weeks, 2 weeks, 30 days, 
- performance status
- 6 months, 8 months
- 18 years
- within 4
- basal cell
- upper limit, limit normal
- life expectancy: life expectancy of at least 3 months
- chemotherapy
- prior studies
- portal vein
- patients must
- patient must
- with the last
- virus
- grade 2
- child pugh: Child-Pugh class A
- ECOG performance
- prior systemic
- nervous system
- locoregional therapy
- ECOG
- 

In [25]:
df.loc[df['eligibility_criteria'].str.contains('child'), 'eligibility_criteria'][22]

'\n        Inclusion Criteria:\n\n          -  HCC diagnosed by positive biopsy or non-invasive criteria,\n\n          -  not suitable for surgical resection or transplantation,\n\n          -  have at least one, but less than or equal to 3 tumors,\n\n          -  of the tumour(s) identified, each tumor must be ≤ 3 cm in diameter,\n\n          -  Child-Pugh class A,\n\n          -  Eastern Cooperative Oncology Group (ECOG) score of 0,\n\n          -  American Society of Anaesthesiologists (ASA) score ≤ 3,\n\n          -  a prothrombin time ratio > 50%,\n\n          -  platelet count > 50x109/L,\n\n          -  ability of patient to stop anticoagulant and anti-platelet therapy for seven days\n             prior to and seven days post NanoKnife procedure,\n\n          -  are able to comprehend and willing to sign the written informed consent form (ICF),\n\n          -  have a life expectancy of at least 3 months.\n\n        Exclusion Criteria:\n\n          -  eligible for surgical treatm